crates/ty_python_semantic/resources/mdtest/type_compendium/tuple.md

```diff
@@ -0,0 +1,122 @@
+# Tuples
+
+## Tuples as product types
+
+Tuples can be used to construct product types. Inhabitants of the type `tuple[P, Q]` are ordered
+pairs `(p, q)` where `p` is an inhabitant of `P` and `q` is an inhabitant of `Q`, analogous to the
+Cartesian product of sets.
+
+```py
+from typing_extensions import assert_type
+
+class P: ...
+class Q: ...
+
+def _(p: P, q: Q):
+    assert_type((p, q), tuple[P, Q])
+```
+
+## Subtyping relationships
+
+The type `tuple[S1, S2]` is a subtype of `tuple[T1, T2]` if and only if `S1` is a subtype of `T1`
+and `S2` is a subtype of `T2`, and similar for other lengths of tuples:
+
+```py
+from ty_extensions import static_assert, is_subtype_of
+
+class T1: ...
+class S1(T1): ...
+class T2: ...
+class S2(T2): ...
+
+static_assert(is_subtype_of(tuple[S1], tuple[T1]))
+static_assert(not is_subtype_of(tuple[T1], tuple[S1]))
+
+static_assert(is_subtype_of(tuple[S1, S2], tuple[T1, T2]))
+static_assert(not is_subtype_of(tuple[T1, S2], tuple[S1, T2]))
+static_assert(not is_subtype_of(tuple[S1, T2], tuple[T1, S2]))
+```
+
+Different-length tuples are not related via subtyping:
+
+```py
+static_assert(not is_subtype_of(tuple[S1], tuple[T1, T2]))
+```
+
+## The empty tuple
+
+The type of the empty tuple `()` is spelled `tuple[()]`. It is *not* a singleton type, as it can be
+subclassed:
+
+```py
+from ty_extensions import static_assert, is_singleton, is_subtype_of, is_equivalent_to, is_assignable_to
+
+static_assert(not is_singleton(tuple[()]))
+
+class AnotherEmptyTuple(tuple[()]): ...
+
+static_assert(not is_equivalent_to(AnotherEmptyTuple, tuple[()]))
+
+# TODO: These should not be errors
+# error: [static-assert-error]
+static_assert(is_subtype_of(AnotherEmptyTuple, tuple[()]))
+# error: [static-assert-error]
+static_assert(is_assignable_to(AnotherEmptyTuple, tuple[()]))
+```
+
+## Singletons?
+
+For the same reason, all other tuples are also not singletons, even if every element is a singleton:
```

Comment:
Yes, it is "for the same reason", but that reason is not subclassing, it's that in the case of non-empty tuples, two tuples with the same contents are not necessarily the same object, not even in CPython as an implementation detail.